### ISSUE TITLE
App from GooglePlay can't be verified

### ISSUE THREAD

**Issue by Giszmo** [btcontract/lnwallet#158]
At the time of writing [this article](https://walletscrutiny.com/posts/2019/11/bictcoinlighningwallet/) we were not able to verify that the app on Google Play matches the published source code. Please fix this for the security of your users.

**Comment by btcontract** [comment]
I use Android Studio 2.3.3 (https://developer.android.com/studio/archive), there is no particular reason I'm not upgrading except that setting up Scala support is a bit tricky with each new release and there are always more urgent tasks at hand.

If  2.3.3 is not an option I do plan to migrate to https://github.com/scala-android/sbt-android soon, will let you know once done.

**Comment by Giszmo** [comment]
May I know if in your team, verification is currently been done, to prevent you from releasing a wallet-stealing version under distress or from a backdoor on your build computer?

**Comment by btcontract** [comment]
No, nothing like that has been done. There's no team, it's a one man shop currently so at least I only have to trust myself with this.

**Comment by Giszmo** [comment]
Well, ... nice you trust yourself but how can users trust that you won't surrender their Bitcoins because of some gun to your head?

**Comment by btcontract** [comment]
Are there any known best practices to deal with this issue?

**Comment by Giszmo** [comment]
Where I am release manager I would do as the robber tells me and rely on my colleagues to catch it before it's too late. If I push to google play without asking for verification, team would sure as hell stop the rollout. I would not worry for the funds of the users.

You being alone could at least make sure outside people could catch it by caring about verifiability. At this stage, our project would list your wallet as verifiable but at a later stage we would also verify your build in real time and alert interested users when updating might be risky.

**Comment by btcontract** [comment]
That's nice, is code ready for this in a current state or should I do something (such as migrate to more recent build tools)?

**Comment by Giszmo** [comment]
You can read my analysis above. I thought building with `./gradlew build` or `./gradlew :app:assembleRelease` or something might work but had no luck. Once you fix that, you can try it out.

Ideally if you follow the build instructions on one machine and then on another and then you compare the two for example with `diffoscope appFromOne.apk appFromTheOther.apk` you should get only a diff in the signature assuming that one apk was done by you and the other by a friend who doesn't have the signing keys.

If there is more than the signature that differs, you might try to investigate more. PNGs might differ due to pngcrunch being non-deterministic. NDK also does cause issues. For those you could try avoiding (pngcrunch you can run once and then disable it) or try running in a very well defined environment. Docker is a great tool for that.

**Comment by Giszmo** [comment]
Any update on this one?
